### DISCUSSION TITLE
Add task to install templates. Similar to RailsBytes

### DISCUSSION THREAD

**Discussion by paulcsmith** [luckyframework/lucky#1286] [Feature requests & Ideas]
Similar to https://railsbytes.com.

For MVP:

* Add `lucky template.install <LINK_TO_TEMPLATE>`. Easy to remember and understand
* A guide on the website for using/creating templates
* To start with people can use GitHub/GitLab/whatever and in the README put a link to the gist/file path
  * Example gist: https://gist.github.com/paulcsmith/4d39b240d16fd6f10d25aa6c66e8e8d5
  * Which can be linked to by clicking "raw" and getting the URL: https://gist.githubusercontent.com/paulcsmith/4d39b240d16fd6f10d25aa6c66e8e8d5/raw/440944bcc711f8d2ebd337f8af2a78653a02ea6d/install_cool_stuff.cr

We should also add some methods like the ones in Thor and Rails templates:

* https://rubydoc.info/github/wycats/thor/master/Thor/Actions
* https://guides.rubyonrails.org/rails_application_templates.html

We don't need them all, but stuff like insert_into_file with the `after` option, creating files, adding shards, asking questions, etc. We have some code in Lucky CLI that does some of this. We just need to move it to something like `LuckyCli::TemplateActions` or something`

## Implementation:

* Takes any URL
* Downloads the contents of the given URL
* Writes the contents to a file somewhere (probably `tasks/temp_lucky_template.cr`), e.g. `template-{random_numbers}.cr`
* The contents should be written with some modifications. It would be wrapped in a Lucky CLI task so it can be run from lucky CLI

```crystal
class InstallTempTemplate < LuckyCli::Task
  include TemplateHelpers # Or whatever we want to call them

  def call
     #{inject_contents_here}
  end
end
```

* Run `lucky install_temp_template` -> this runs all the code from the linked URL, which now has access to LuckyCLI so it can use all the helper methods we define there
* Wrap whole thing in `begin/ensure` and an `at_exit` so that if something fails we can print an error message *and* delete the file from src/tasks

## Security

* The file can run any Crystal code, so could do all kinds of malicious stuff. Then again, this is true of any shard you install. So we should recommend to always look at the file first before running the template command

## Later on

We should add something to the website to explore templates. This would be super simpler. It doesn't have a code editor or anything like that. You can sign up, post a "template" with:

* Name
* Short description
* Usage instructions
* Link to template

So the website is just a place for people to share templates and make them easier to find and install

Thoughts?

**Comment by stephendolan** [comment]
Love it! I'd actually recommend that we emulate tty-file over thor, though, having had a much easier time digging into TTY's code in the past: https://github.com/piotrmurach/tty-file

**Reply by paulcsmith** [reply]
That sounds good to me. I was mostly thinking the external API like reading files, and appending text. Implementation can def be different though!

**Comment by jkthorne** [comment]
I have given this some thought because I am not sure it templates are the best solution but I cannot think of better way to do this.  Another thought is something like [Ameba](https://github.com/crystal-ameba/ameba).  An in place replacement is easier in Crystal because Crystal allows access to the code at compile time.  The pro is a higher level of confidence in applying a template. The cons is it is way harder to make template and it would really only be able to modify Crystal files.

**Comment by stephendolan** [comment]
I've been working on a solution that we could eventually use for this, and it's pretty nifty!

https://github.com/stephendolan/spark

Still a work in progress, but I've really enjoyed using it with Lucky Jumpstart's `init` task so far.

**Reply by stephendolan** [reply]
As of [v1.14.0](https://github.com/stephendolan/spark/releases/tag/v1.14.0), Spark now has support for template application. Adding a task like Paul originally submitted in this discussion is trivial, but I'm going to play with it for a while in my Lucky Jumpstart repository first!
